### ISSUE TITLE
operator: pod store is never initialized

### ISSUE THREAD

**Issue by vflaux** [cilium/cilium#32713]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Each time a node needs more IP, the operator will log a warning indicating that the pod store is not initialized.
As a result the surge allocation is skipped and the operator will not allocate more than the pre-allocate value for the node (defaults to 8).

### Cilium Version

1.15.0

### Kernel Version

Linux xxx 5.10.215-203.850.amzn2.x86_64 #1 SMP Tue Apr 23 20:32:19 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

v1.28.8-eks-adc7111

### Regression

https://github.com/cilium/cilium/commit/1a9a1e7b2a7bebdf4a87200c994af91b58dba097#diff-feb41af31e6ee7149118809ccc6fe42b0603dbb5e8f9c16a72bc2fe492c0b921L56-L61

### Relevant log output

```shell
level=warning msg="Unable to compute pending pods, will not surge-allocate" error="pod store uninitialized" instanceID=i-xxx name=xxx subsys=ipam
level=info msg="Resolving IP deficit of node" available=16 availableForAllocation=208 emptyInterfaceSlots=3 instanceID=i-xxx maxIPsToAllocate=1 name=xxx neededIPs=1 remainingInterfaces=4 selectedInterface=eni-xxx selectedPoolID=subnet-xxx subsys=ipam used=9
```

### Anything else?

The pod store in `operator/watchers/pod.go` is created by the function `PodsInit` which is never called anywhere in the code.

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by joestringer** [comment]
Hi, can you confirm whether you experience this with the latest v1.15.x release? I see the same log mentioned in https://github.com/cilium/cilium/pull/31016 , so I'm not sure whether it was already resolved in a newer Cilium bugfix release.

**Comment by vflaux** [comment]
Hi, I could, but the code hasn't changed in the `v1.15` branch or in `main`, so I'm pretty sure the result is the same.
You should see this log systematically since Cilium 1.15.

**Comment by joestringer** [comment]
OK, thanks for confirming. It looks like `PodsInit` was last in the code prior to 1a9a1e7b2a7bebdf4a87200c994af91b58dba097 (https://github.com/cilium/cilium/commit/1a9a1e7b2a7bebdf4a87200c994af91b58dba097). I acknowledge it's no longer called in any v1.15.x released versions. @alan-kut do you know why this path is no longer called after the migration to Hive? Is this expected to be implicitly called somehow or was this an oversight?

**Comment by alex-berger** [comment]
Observing the same behavior with Cilium version 1.15.6. What negative effect does this have, what will not work if this happens?

**Comment by vflaux** [comment]
tl;dr: it just delays the allocation of IPs when a node suddenly need a lot of IPs (16 for example).

If a node need more IP than what have been pre allocated (8 by default), it will have to wait for the next reconciliation cycle (each minutes) for the cilium operator to try to allocate more IPs. In my understanding of the code, this is the expected behavior so far.
When the reconciliation occurs, the controller tries to list the pending pods for the node to know how many IPS it needs to allocate (called "surge allocation" in the code). Without it, the controller just ask to allocates the number of IPs needed for the node to have its number of "pre allocated" satisfied. But this is only based on the number of IP already allocated and actually in use. As a result it will never ask to allocate more than this "pre-allocate value" per reconciliation cycle (when "allocated" equals "in use").
This is not a problem until it needs more than 2x the "pre-allocate value". If you need more, it will just delay the allocation of the next packet of IPs until the next reconciliation (one minute).
It most likely do not affect many real world clusters, and that could explain why this has not been reported since the 1.15.0.
I found this issue in a testing environment on eks by allocating 80 pods to a node instantly. Cilium took minutes to allocate all the IPs needed which surprised me.

**Comment by alex-berger** [comment]
@vflaux Thank you very much for that explanation, that saved me several hours or days of reading through the code :-)

I know "one minute" might not sound much, but in our case where we are (beside others) running GitLab Runners (CI Jobs) on Kubernetes, each second matters. Our clusters are optimized for fast Pod startup and fully automatic scaling using tools like [Karpenter](https://karpenter.sh/). So, fixing issues like this will make the difference :-)

**Comment by Dhouti** [comment]
Also encountering this issue. Observed while doing a rollout of a new AMI in EKS, some pods being scheduled to a new node would go into crashloopbackoff for several minutes due to lack of IPs until the operator caught up. 

Looking at enabling prefix delegation, will likely function as a workaround for this.

**Comment by alex-berger** [comment]
I was somehow hoping, that this might get fixed for 1.16.0, but it was not fixed yet.

@joestringer Are there any plans to fix this?

**Comment by bowei** [comment]
@dlapcevic
